### PR TITLE
Add a Dockerfile

### PR DESCRIPTION
**Pull request by davidmehren**
### Component/Part
Docker

### Description
This PR adds a `Dockerfile` and necessary changes to our Yarn infrastructure to install prod-only dependencies.
It also includes a refactor that ensures the app works with prod-only dependencies.

I added the documentation for using the Dockerfile directly in the `docker` directory, instead of the docs, as I found it more immediately helpful there. Especially as docs.hedgedoc.org is deployed from `master`.

I plan to add more infrastructure to the `docker` folder in the future, including a `docker-compose.yml` to help testing with different databases.

### Steps

- [x] Added implementation
- [x] Added / updated documentation
- [x] I read the [contribution documentation](https://github.com/hedgedoc/hedgedoc/blob/develop/CONTRIBUTING.md) and
  made sure that:
  - My commits are signed-off to accept the DCO.
  - This PR targets the correct branch: `master` for 1.x & docs, `develop` for 2.x

### CODE REVIEW

**Comment by DerMolly** [comment]
One thought that just came up: Do we want to keep the `.spec`-Files in the container? I mean we need them in the development stage, but strictly speaking leaving them out for the production container would save a bit space.

File: docker/README.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+## Build a production image
+**Note:** This does not include any frontend!
+
+To build a production image, run the following command from the root of the repository:  
+`docker build -t hedgedoc-prod -f docker/Dockerfile .`
+
+When you run the image, you need to provide environment variables to configure HedgeDoc.
+This example starts HedgeDoc on localhost, with non-persistent storage:  
+`docker run -e HD_DOMAIN=localhost -e HD_MEDIA_BACKEND=filesystem -e HD_MEDIA_BACKEND_FILESYSTEM_UPLOAD_PATH=uploads -e HD_DATABASE_DIALECT=sqlite -e HD_DATABASE_STORAGE=hedgedoc.sqlite -e HD_SESSION_SECRET=foobar -e HD_LOGLEVEL=debug -p 3000:3000 hedgedoc-prod`
```
Comment (DerMolly) [inline docker/README.md:23]:
```suggestion
`docker run -e HD_DOMAIN=http://localhost -e HD_MEDIA_BACKEND=filesystem -e HD_MEDIA_BACKEND_FILESYSTEM_UPLOAD_PATH=uploads -e HD_DATABASE_DIALECT=sqlite -e HD_DATABASE_STORAGE=hedgedoc.sqlite -e HD_SESSION_SECRET=foobar -e HD_LOGLEVEL=debug -p 3000:3000 hedgedoc-prod`
```

File: docker/README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+The `Dockerfile` in this repo uses multiple stages and can be used to create both images for development
+and images with only production dependencies.
+It uses features which are only available in BuildKit - see https://docs.docker.com/go/buildkit/ for more information.
+
+## Build a production image
+**Note:** This does not include any frontend!
+
+To build a production image, run the following command from the root of the repository:  
+`docker build -t hedgedoc-prod -f docker/Dockerfile .`
```
[the remark this answers is not in the document]
Comment (davidmehren) [reply docker/README.md:19]: I will emphasize the note about the working directory. Unfortunately, due to the way docker works, there is no way around this.

File: docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+FROM base as prod
+
+USER node
+WORKDIR /usr/src/app
+ENV NODE_ENV=production
+
+COPY --chown=node --from=builder /usr/src/app/dist ./
+
+RUN --mount=type=cache,sharing=locked,uid=1000,gid=1000,target=/tmp/.yarn \
+    YARN_CACHE_FOLDER=/tmp/.yarn yarn install --immutable && \
```
Comment (ErikMichelson) [inline docker/Dockerfile:68]: This install is superfluous as the `yarn workspaces focus` already does an install. See [the docs](https://yarnpkg.com/cli/workspaces/focus). So this install just adds all development dependencies as well (which is not what we want in the prod image).

**Comment by ErikMichelson** [comment]
We could add the [OCI Annotations](https://github.com/opencontainers/image-spec/blob/main/annotations.md) to the resulting docker image.

**Review by ErikMichelson — CHANGES_REQUESTED** [review]
## Production image

This image does not work!
The copy of the `dist` folder directly into the root of `/usr/src/app` results in invalid path levels, leading to errors like this one:
```
[HedgeDoc.ExceptionsHandler()] ENOENT: no such file or directory, open '/usr/src/package.json'
```

Otherwise, the production image is good, especially the size is a decrese of 150 MB compared to the alpine docker image of HedgeDoc 1.x.

<details>
<summary>Stats</summary>

```
Total Image size: 234 MB
Potential wasted space: 730 kB
Image efficiency score: 99 %
```

</details>

## Development image

The development image works as expected.

<details>
<summary>Stats</summary>

```
Total Image size: 459 MB
Potential wasted space: 730 kB
Image efficiency score: 99 %
```

</details>

**Review by DerMolly — APPROVED** [review]

**Review by InnayTool — APPROVED** [review]

**Review by ErikMichelson — APPROVED** [review]

**Pull request merged**